File: web/src/utils/phraseEditing.js
import { createPhrase } from '../models/song.js';

function separatorAt(separators, measure) {
    return [...(separators || [])]
        .filter(separator => separator.fromMeasure <= measure)
        .sort((a, b) => b.fromMeasure - a.fromMeasure)[0] || null;
}

export function splitPhraseAtMeasure(phrase, splitMeasure, unitsPerMeasure, newName) {
    if (!Number.isInteger(splitMeasure) || splitMeasure <= 0 || splitMeasure >= phrase.length) return null;
    const splitTime = splitMeasure * unitsPerMeasure;
    const partition = notes => ({
        before: notes.filter(note => note.startTime < splitTime),
        after: notes.filter(note => note.startTime >= splitTime)
            .map(note => ({ ...note, startTime: note.startTime - splitTime })),
    });
    const melody = partition(phrase.tracks.melody);
    const chords = partition(phrase.tracks.chords);
    const beforeSeparators = (phrase.handSeparators || []).filter(s => s.fromMeasure < splitMeasure);
    const afterSeparators = (phrase.handSeparators || [])
        .filter(s => s.fromMeasure >= splitMeasure)
        .map(s => ({ ...s, fromMeasure: s.fromMeasure - splitMeasure }));
    const inherited = separatorAt(phrase.handSeparators, splitMeasure);
    if (inherited && !afterSeparators.some(s => s.fromMeasure === 0)) {
        afterSeparators.unshift({ ...inherited, fromMeasure: 0 });
    }

    const after = createPhrase(newName, phrase.length - splitMeasure);
    after.tracks = { melody: melody.after, chords: chords.after };
    after.handSeparators = afterSeparators;
    return [
        {
            ...phrase,
            length: splitMeasure,
            tracks: { melody: melody.before, chords: chords.before },
            handSeparators: beforeSeparators,
        },
        after,
    ];
}

export function mergePhrases(previous, current, unitsPerMeasure) {
    const offset = previous.length * unitsPerMeasure;
    const shiftedSeparators = (current.handSeparators || []).map(separator => ({
        ...separator,
        fromMeasure: separator.fromMeasure + previous.length,
    }));
    return {
        ...previous,
        // Phrase length is authored structure, including intentional trailing silence.
        length: previous.length + current.length,
        tracks: {
            melody: previous.tracks.melody.concat(current.tracks.melody.map(note => ({ ...note, startTime: note.startTime + offset }))),
            chords: previous.tracks.chords.concat(current.tracks.chords.map(note => ({ ...note, startTime: note.startTime + offset }))),
        },
        handSeparators: [...(previous.handSeparators || []), ...shiftedSeparators],
    };
}

File: android/app/src/main/kotlin/com/tobietheunknown/pianoteacher/utils/PhraseEditing.kt
package com.tobietheunknown.pianoteacher.utils

import com.tobietheunknown.pianoteacher.data.model.*
import java.util.UUID

fun splitPhraseAtMeasure(
    phrase: Phrase,
    splitMeasure: Int,
    unitsPerMeasure: Double,
    newName: String = "${phrase.name} bis",
): Pair<Phrase, Phrase>? {
    if (splitMeasure <= 0 || splitMeasure >= phrase.length || !unitsPerMeasure.isFinite() || unitsPerMeasure <= 0) return null
    val splitTime = splitMeasure * unitsPerMeasure
    // Keep crossing holds with their attack: splitting must not add a reattack
    // or shorten a note when playing the complete song or merging it again.
    fun partition(notes: List<NoteEvent>) = notes.filter { it.startTime < splitTime } to
        notes.filter { it.startTime >= splitTime }.map { it.copy(startTime = it.startTime - splitTime) }
    val (melodyBefore, melodyAfter) = partition(phrase.tracks.melody)
    val (chordsBefore, chordsAfter) = partition(phrase.tracks.chords)
    val beforeSeparators = phrase.handSeparators.filter { it.fromMeasure < splitMeasure }
    val afterSeparators = phrase.handSeparators.filter { it.fromMeasure >= splitMeasure }
        .map { it.copy(fromMeasure = it.fromMeasure - splitMeasure) }.toMutableList()
    val inherited = phrase.handSeparators.filter { it.fromMeasure <= splitMeasure }.maxByOrNull { it.fromMeasure }
    if (inherited != null && afterSeparators.none { it.fromMeasure == 0 }) {
        afterSeparators.add(0, inherited.copy(fromMeasure = 0))
    }
    return phrase.copy(
        length = splitMeasure,
        tracks = Tracks(melodyBefore, chordsBefore),
        handSeparators = beforeSeparators,
    ) to phrase.copy(
        id = UUID.randomUUID().toString(),
        name = newName,
        length = phrase.length - splitMeasure,
        tracks = Tracks(melodyAfter, chordsAfter),
        handSeparators = afterSeparators,
    )
}

fun mergePhrases(previous: Phrase, current: Phrase, unitsPerMeasure: Double): Phrase {
    val offset = previous.length * unitsPerMeasure
    return previous.copy(
        length = previous.length + current.length,
        tracks = Tracks(
            previous.tracks.melody + current.tracks.melody.map { it.copy(startTime = it.startTime + offset) },
            previous.tracks.chords + current.tracks.chords.map { it.copy(startTime = it.startTime + offset) },
        ),
        handSeparators = previous.handSeparators + current.handSeparators.map {
            it.copy(fromMeasure = it.fromMeasure + previous.length)
        },
    )
}
